packages/misc/src/lib/misc.ts

```diff
@@ -691,3 +694,113 @@ export function getEnv({
   // Default
   return defaultValue;
 }
+
+export function sendRequest(
+  url: string,
+  req: RequestInit,
+  requestId: string
+): Promise<Response> {
+  return fetch(url, req)
+    .then(async (response) => {
+      const isJson = response.headers
+        .get('content-type')
+        ?.includes('application/json');
+
+      const data = isJson ? await response.json() : null;
+
+      if (!response.ok) {
+        // get error message from body or default to response status
+        const error = data || response.status;
+        return Promise.reject(error);
+      }
+
+      return data;
+    })
+    .catch((error: NodeErrorV3) => {
+      logErrorWithRequestId(
+        requestId,
+        `Something went wrong, internal id for request: lit_${requestId}. Please provide this identifier with any support requests. ${
+          error?.message || error?.details
+            ? `Error is ${error.message} - ${error.details}`
+            : ''
+        }`
+      );
+      return Promise.reject(error);
+    });
+}
+
+export function executeWithRetry(
+  execCallback: (
+    requestId: string
+  ) => Promise<SuccessNodePromises<any> | RejectedNodePromises>,
+  errorCallback?: (error: any, requestId: string, isFinal: boolean) => void,
+  opts?: any
+): Promise<
+  | (SuccessNodePromises<any> & { requestId: string })
+  | (RejectedNodePromises & { requestId: string })
+> {
+  let timer: any | null;
+  let counter = 0;
+  let isTimeout = false;
+  const poll = async (): Promise<any> => {
+    if (!opts) {
+      opts = {};
+    }
+    opts.timeout = opts.timeout ?? 31_000; // We wait for 31 seconds as the timeout period on the nodes is 30 seconds.
+    opts.interval = opts.interval ?? 100;
+    opts.maxRetryCount = opts.maxRetryCount ?? 3;
+    let requestId: string;
+    console.log('starting nettwork operation');
+    while (!isTimeout) {
+      requestId = Math.random().toString(16).slice(2);
+      try {
+        timer = setTimeout(() => {
+          isTimeout = true;
+        }, opts.timeout);
+        const response: any = await execCallback(requestId).catch((err) => {
+          counter += 1;
+          errorCallback &&
+            errorCallback(
+              `Error is ${err.message}-${err.details}`,
+              requestId,
+              counter >= opts.maxRetryCount ? true : false
+            );
+        });
+
+        clearTimeout(timer);
+        response.requestId = requestId;
+        if ((response as RejectedNodePromises).error) {
```

Comment:
yes it will `return` and error since the callers of this method expect an error response. we do not throw execptions here as to let callers process the error object and then do something with it. since this is a low level utility its trying to be as generic with returning data to the caller as possible hence using the `succces` and `failure` types